lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,746 @@
+    },
+
+    abort(reason) {
+      if (!streamWritable.destroyed && !errored)
+        streamWritable.destroy(reason);
+    },
+
+    close() {
+      if (!streamWritable.writableEnded && !closed)
+        streamWritable.end();
+    },
+  }, strategy);
+
+  return stream;
+}
+
+function newReadableStreamFromStreamReadable(streamReadable) {
+  if (typeof streamReadable?._readableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamReadable',
+      'stream.Readable',
+      streamReadable);
+  }
+
+  const objectMode = streamReadable.readableObjectMode;
+  const highWaterMark = streamReadable.readableHighWaterMark;
+  const strategy =
+    objectMode ?
+      new CountQueuingStrategy({ highWaterMark }) :
+      { highWaterMark };
+
+  const type = !objectMode ? 'bytes' : undefined;
+  const autoAllocateChunkSize = !objectMode ? highWaterMark : undefined;
+
+  let controller;
+  let closed = false;
+
+  function onClose() {
+    closed = true;
+    streamReadable.off('close', onClose);
+    streamReadable.off('data', onData);
+    streamReadable.off('end', onEnd);
+    streamReadable.off('error', onError);
+  }
+
+  function onError(error) {
+    controller.error(error);
+  }
+
+  function onData(chunk) {
+    // Copy the Buffer to detach it from the pool.
+    if (Buffer.isBuffer(chunk))
+      chunk = new Uint8Array(chunk);
+    controller.enqueue(chunk);
+  }
+
+  function onEnd() {
+    controller.close();
+  }
+
+  streamReadable.once('close', onClose);
+  streamReadable.once('end', onEnd);
+  streamReadable.once('error', onError);
+  streamReadable.on('data', onData);
+  streamReadable.pause();
+
+  return new ReadableStream({
+    type,
+    autoAllocateChunkSize,
+
+    start(c) { controller = c; },
+
+    pull() { streamReadable.resume(); },
+
+    cancel(reason) {
+      if (!closed)
+        streamReadable.destroy(reason);
+    },
+  }, strategy);
+}
+
+function newReadableWritablePairFromDuplex(duplex) {
+  if (typeof duplex?._writableState !== 'object' ||
+      typeof duplex?._readableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE('duplex', 'stream.Duplex', duplex);
+  }
+
+  const writable = newWritableStreamFromStreamBase(duplex);
+  const readable = newReadableStreamFromStreamBase(duplex);
+
+  return {
+    writable,
+    readable,
+  };
+}
+
+function newStreamWritableFromWritableStream(writableStream, options = {}) {
+  if (!isWritableStream(writableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'writableStream',
+      'WritableStream',
+      writableStream);
+  }
+
+  validateObject(options, 'options');
+  const {
+    objectMode = false,
+    highWaterMark,
+    signal,
+  } = options;
+
+  validateBoolean(objectMode, 'options.objectMode');
+  if (signal !== undefined)
+    validateAbortSignal(signal, 'options.signal');
+
+  const writer = writableStream.getWriter();
+  let closed = false;
+
+  const writable = new Writable({
+    highWaterMark: highWaterMark || writer.desiredSize,
```

Comment:
This seems weird. If the `WritableStream` has a HWM of e.g. 1 MB and no explicit HWM was specified in `options`, should the `Writable` then **also** have a HWM of 1 MB? That means we'd be have a total buffer of up to **2 MB** (instead of 1 MB).